src/v/storage/disk_log_impl.cc

```diff
@@ -507,40 +507,47 @@ segment_set disk_log_impl::find_sliding_range(
             // Stop once we get to an unstable segment.
             break;
         }
+        if (
+           _last_compaction_window_start_offset.has_value()
+           && (seg->offsets().get_base_offset()
+                >= _last_compaction_window_start_offset.value())) {
+            // Force clean segment production by compacting down to the
+            // start of the log before considering new segments in the
+            // compaction window.
+            break;
+        }
         if (
           new_start_offset
           && seg->offsets().get_base_offset() < new_start_offset.value()) {
             // Skip over segments that are being truncated.
             continue;
         }
+
         buf.emplace_back(seg);
     }
     segment_set segs(std::move(buf));
     if (segs.empty()) {
         return segs;
     }
 
-    // If a previous sliding window compaction ran, and there are no new
-    // segments, segments at the start of that window and above have been
-    // fully deduplicated and don't need to be compacted further.
-    //
-    // If there are new segments that have not been compacted, we can't make
-    // this claim, and compact everything again.
-    if (
-      segs.back()->finished_windowed_compaction()
-      && _last_compaction_window_start_offset.has_value()) {
-        while (!segs.empty()) {
-            if (
-              segs.back()->offsets().get_base_offset()
-              >= _last_compaction_window_start_offset.value()) {
-                // A previous compaction deduplicated the keys above this
-                // offset. As such, segments above this point would not benefit
-                // from being included in the compaction window.
-                segs.pop_back();
-            } else {
-                break;
+    // If all of the segments in the set have already been made clean by window
+    // compaction, it's likely because the _last_compaction_window_start_offset
+    // has been reset after compacting all the way down to the start of the log
+    // (and no segments have been added/closed since). This would result in a
+    // no-op to compact. The exception to this is segments with removable
+    // tombstones. We can allow for frequent tombstone removal by leaving them
+    // in the compaction window.
+    const bool all_clean_compacted = std::ranges::all_of(
+      segs,
+      [](const auto& s) { return s->index().has_clean_compact_timestamp(); });
+    if (all_clean_compacted) {
+        segment_set::underlying_t filtered_buf;
```

Comment:
Really dislike this manual method of filtering, would much prefer STL use or something more efficient